File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int i = 1, j = height.size()-2;
        int lmax = height[0], rmax = height[height.size()-1];
        int total_water = 0;
        while (i <= j){
            int water = 0;
            if (lmax >= rmax){
                water = max(0, rmax - height[j]);
                rmax = max(rmax, height[j]);
                j--;
            } else {
                water = max(0, lmax - height[i]);
                lmax = max(lmax, height[i]);
                i++;
            }
            total_water += water;
        }
        return total_water;
    }
};

File: 42-trapping-rain-water/trapping-rain-water.go
func trap(height []int) int {
    total_water := 0
    l, r := 1, len(height)-2
    l_max, r_max := height[l-1], height[r+1]

    for l <= r {
        water := 0

        if l_max > r_max {
            water = max(0, (r_max - height[r]))
            r_max = max(r_max, height[r])
            r--
        } else {
            water = max(0, (l_max - height[l]))
            l_max = max(l_max, height[l])
            l++
        }
        total_water += water
    }

    return total_water
}